File: Dynamic_Programming/714.Best-Time-to-Buy-and-Sell-Stock-with-Transaction-Fee/714.Best-Time-to-Buy-and-Sell-Stock-with-Transaction-Fee.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) 
    {
        if (prices.size()==1) return 0;
        
        int s0=0;
        int s1=-prices[0];
        
        for (int i=1; i<prices.size(); i++)
        {
            int temp = s0;
            s0 = max(s0, s1+prices[i]-fee);
            s1 = max(temp-prices[i],s1);
        }                
        
        return s0;
    }
};

File: Dynamic_Programming/714.Best-Time-to-Buy-and-Sell-Stock-with-Transaction-Fee/714.Best-Time-to-Buy-and-Sell-Stock-with-Transaction-Fee.py
class Solution:
    def maxProfit(self, prices, fee):
        """
        :type prices: List[int]
        :type fee: int
        :rtype: int
        """
        hold, sold = -math.inf,0
        for p in prices:
            hold_tmp, sold_tmp = hold, sold
            hold = max(hold_tmp, sold_tmp-p)
            sold = max(sold_tmp, hold_tmp+p-fee)
        return max(hold,sold)
